### PR TITLE
BUG: read empty files correctly

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes https://github.com/geopandas/geopandas/issues/649

### CODE REVIEW

**Comment by WeatherGod** [comment]
First, you should add a unit test to make sure this doesn't break in the future (feel free to use the zipped shapefile I attached in that issue). Second, if I am reading this right, the dtypes aren't getting set, and would fall back to some default that has nothing to do with the schema of the shapefile?

File: geopandas/io/file.py [hunk 1]
```diff
@@ -67,9 +67,14 @@ def read_file(filename, **kwargs):
         else:
             f_filt = f
         gdf = GeoDataFrame.from_features(f_filt, crs=crs)
-        # re-order with column order from metadata, with geometry last
+
         columns = list(f.meta["schema"]["properties"]) + ["geometry"]
-        gdf = gdf[columns]
+        if gdf.empty:
+            # make sure the empty result has the correct column names
+            gdf = GeoDataFrame(columns=columns)
```
Comment (WeatherGod) [inline geopandas/io/file.py]: Isn't the bug really in `from_features()`? Since it is the method that was returning an empty geodataframe?

**Comment by WeatherGod** [comment]
Ah, I missed the unit test. But I am not convinced this is the right solution. And the unit test doesn't confirm the dtypes.

File: geopandas/io/file.py (hunk 1 above)
Comment (jorisvandenbossche) [reply geopandas/io/file.py]: Yes, and I first wanted to fix it there. However, the problem is that `from_features` is more general, and does not only handle fiona collections. A general FeatureCollection (eg from `__geo_interace__`) has no information about the columns if it is empty.

**Comment by jorisvandenbossche** [comment]
OK, changed it a bit to let `from_features` take a `columns` keyword that ensures the output.

You are right that this does not yet preserve the type information from the shapefile. This is not something that can be solved itself in `from_features`, but can in principle certainly be done inside `read_file` based on the metadata of the fiona collection. We would need to do some mapping of the types. 
I am going to leave this PR for just the empty fix (it's already a net improvement that it doesn't raise anymore and keeps the column names), but a PR to also fix the dtypes is certainly welcome!

**Comment by jorisvandenbossche** [comment]
@jdmcbr any comments?

File: geopandas/io/tests/test_io.py [hunk 2]
```diff
@@ -70,3 +72,24 @@ def test_filtered_read_file(self):
         filtered_df_shape = filtered_df.shape
         assert full_df_shape != filtered_df_shape
         assert filtered_df_shape == (2, 5)
+
+    def test_emtpy_shapefile(self, tmpdir):
```
Comment (jdmcbr) [inline geopandas/io/tests/test_io.py]: `emtpy` to `empty`

File: geopandas/io/file.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
 import fiona
 import numpy as np
 import six
+from shapely.geometry import shape
```
Comment (jdmcbr) [inline geopandas/io/file.py]: Unless I'm missing something, this import doesn't look necessary.

Comment (jorisvandenbossche) [reply geopandas/io/file.py]: yes, thanks for noticing. I had an intermediate version that needed this :-)

**Pull request merged**